### PR TITLE
Made format look good on Slack and Teams

### PR DESCRIPTION
**Pull request by subashc2023**
image

image

### CODE REVIEW

File: get_forecast.py [hunk 1]
```diff
@@ -324,60 +324,57 @@ def calc_forecast(boto3_session):
...
-    for line in formated_rows :
-        formated_line=""
-        for column in columns_displayed :
-            if formated_line != "" :
-                formated_line += " "
-            formated_line += line.get(column)
-        message += formated_line.rstrip() + "\n"
+    message = ""
+    for row in formated_rows:
+        message += row['Account'] + " | " + row['Forecast'] + " | " + row['Change'] + "\n"
```
Comment (jimzucker) [inline get_forecast.py]: This change is not correct you have lost the column display configuration feature that the code you removed handled.  You correctly removed it from the default list above but should not have remove the feature to add it back via configuration without discussing it.  this kind of side effect you have to watchout for.

File: get_forecast.py [hunk 2]
```diff
@@ -324,60 +324,57 @@ def calc_forecast(boto3_session):
...
+    # print in descending order by forecast
     lines = sorted(output, key=lambda k: k.get('amount_forecast'), reverse=True)
-    for line in lines :
+    for line in lines:
         if len(lines) == 2 and line.get('account_name') == 'Total':
             continue
         change = "{0:,.1f}%".format(line.get('forecast_variance'))
         row = {
             "Account": line.get('account_name')[:account_width].ljust(account_width),
-            "MTD": "${0:,.0f}".format(line.get('amount_usage')).rjust(mtd_width),
```
Comment (jimzucker) [inline get_forecast.py:351]: MTD shoul dnot ve removed from the row, if you look at how the code is written it is configurable, all you had to dos was suppress it on. line 361 and maybe improve the construction at the end but not sure

File: get_forecast.py [hunk 3]
```diff
@@ -324,60 +324,57 @@ def calc_forecast(boto3_session):
...
 
-    output_rows=[]
+    total_forecast = 0
+    total_change = 0
+    
+    output_rows = []
 
-    row = {
-        "Account": 'Account'.ljust(account_width),
-        "MTD": 'MTD'.rjust(mtd_width),
```
Comment (jimzucker) [inline get_forecast.py:337]: incorrect to remoe MTD here, you should only modify the display the code is designed for it to be configurabel.  before making a change to remove the feature you should discuss.. you simply supresss in line 361

File: get_forecast.py [hunk 4]
```diff
@@ -324,60 +324,57 @@ def calc_forecast(boto3_session):
     return output
 
 
-def format_rows(output,account_width):
-    #print the heading
-    mtd_width=8
```
Comment (jimzucker) [inline get_forecast.py:329]: incorrect to remove

File: get_forecast.py [hunk 5]
```diff
@@ -109,7 +109,7 @@ def send_teams(teams_url, message):
         return
 
     teams_message = {
-        'text': message
+        'text': '```\n' + message + '```\n'
```
Comment (jimzucker) [inline get_forecast.py]: How do these differ that we need 2 of them with duplicate code?

File: get_forecast.py [hunk 6]
```diff
@@ -86,7 +86,7 @@ def send_slack(slack_url, message):
         return
 
     slack_message = {
-        'text': message
+        'text': '```\n' + message + '```\n'
```
Comment (jimzucker) [inline get_forecast.py]: How do these differ that we need 2 of them with duplicate code?

File: get_forecast.py [hunk 7]
```diff
@@ -324,60 +324,57 @@ def calc_forecast(boto3_session):
...
-    forecast_width=8
-    change_width=6
+def format_rows(output, account_width):
+    # print the heading
+    forecast_width = 8
+    change_width = 8
 
-    output_rows=[]
+    total_forecast = 0
+    total_change = 0
```
Comment (jimzucker) [inline get_forecast.py]: I dont see this used anywhere

File: get_forecast.py [hunk 8]
```diff
@@ -324,60 +324,57 @@ def calc_forecast(boto3_session):
...
-    mtd_width=8
-    forecast_width=8
-    change_width=6
+def format_rows(output, account_width):
+    # print the heading
+    forecast_width = 8
+    change_width = 8
 
-    output_rows=[]
+    total_forecast = 0
```
Comment (jimzucker) [inline get_forecast.py]: I dont see this used anywhere

**Review by jimzucker — CHANGES_REQUESTED** [review]
Please see comment address and mark each point resolved and then I will re-review..  LMK

removing MTD should have only been to change the conf on I think line 361 and maybe you needed to ehnace the loop that decides what to put in the output row or not, not sure..

**Comment by subashc2023** [comment]
image
With and without MTD on Slack

**Comment by subashc2023** [comment]
image

File: get_forecast.py [hunk 9]
```diff
@@ -324,60 +324,65 @@ def calc_forecast(boto3_session):
     return output
 
 
-def format_rows(output,account_width):
-    #print the heading
+def format_rows(output, account_width):
+    # print the heading
     mtd_width=8
```
Comment (jimzucker) [inline get_forecast.py]: you are reformatting xx=yy to xx = yy you have to be consistent...

File: get_forecast.py (hunk 6 above)
Comment (jimzucker) [reply get_forecast.py]: For not a better solution would be to make a constant

code_block_format_message= '```\n' + message + '```\n'

and use the constant in 2 places

**Review by jimzucker — DISMISSED** [review]

File: get_forecast.py (hunk 6 above)
Comment (jimzucker) [reply get_forecast.py]: I code this

code_block_format_mssage= '```\n' + message + '```\n'

slack_message = {
         'text':  code_block_format_mssage
}

teams_message = {
         'text':  code_block_format_mssage
}

**Comment by subashc2023** [comment]
image

**Comment by subashc2023** [comment]
image

File: get_forecast.py [hunk 10]
```diff
@@ -80,13 +80,14 @@ def get_secret(sm_client):
     return secret
 
 
-def send_slack(slack_url, message):
+def send_slack(slack_url, code_block_format_message):
```
Comment (jimzucker) [inline get_forecast.py]: Passing it is is a great idea but in that case the variable name does not make sense, this opera should be text_message as the caller will decide if it si code_block or something else

File: get_forecast.py [hunk 11]
```diff
@@ -103,13 +104,13 @@ def send_slack(slack_url, message):
         logger.error("slack_url= %s", slack_url)
         raise e
   
-def send_teams(teams_url, message):
+def send_teams(teams_url, code_block_format_message):
```
Comment (jimzucker) [inline get_forecast.py]: same

File: get_forecast.py [hunk 12]
```diff
@@ -325,60 +326,63 @@ def calc_forecast(boto3_session):
...
+    for line in formated_rows:
+        formated_line = ""
+        for column in columns_displayed:
+            if formated_line != "":
+                formated_line += " | "
             formated_line += line.get(column)
         message += formated_line.rstrip() + "\n"
-    display_output(boto3_session, message)
+
+    code_block_format_message = '```\n' + message + '```\n'
```
Comment (jimzucker) [inline get_forecast.py:384]: this is a good change

**Review by jimzucker — CHANGES_REQUESTED** [review]

**Comment by subashc2023** [comment]
image

**Review by jimzucker — APPROVED** [review]
LTGM - Good catch on the default in the readme

**Pull request merged**